### PR TITLE
New API for extended monitoring failure statuses

### PR DESCRIPTION
**Pull request by andrey-mysterium**

### CODE REVIEW

File: core/node/node_monitoring_agent_tracker.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package node
+
+// MonitoringAgentStatuses a object represent a [service_type][status]amount of statuses for each status type
```
Comment (soffokl) [inline core/node/node_monitoring_agent_tracker.go:20]:
```suggestion
// MonitoringAgentStatuses a object represent a [service_type][status]amount of statuses for each service type.
```

File: core/quality/mysterium_morqa.go [hunk 2]
```diff
@@ -288,6 +289,32 @@ func (m *MysteriumMORQA) ProviderSessions(providerID string) []ProviderSession {
 	return responseBody.Connects
 }
 
+// ProviderStatuses fetch provider connectivity statuses from prometheus (monitoring agent)
```
Comment (soffokl) [inline core/quality/mysterium_morqa.go:292]:
```suggestion
// ProviderStatuses fetch provider connectivity statuses from quality oracle.
```

File: core/quality/mysterium_morqa.go [hunk 3]
```diff
@@ -288,6 +289,32 @@ func (m *MysteriumMORQA) ProviderSessions(providerID string) []ProviderSession {
...
+	request, err := m.newRequestJSON(http.MethodGet, fmt.Sprintf("providers/statuses?provider_id=%s", providerID), "")
+	if err != nil {
+		log.Warn().Err(err).Msg("Failed to create provider monitoring agent statuses request")
+		return nil
+	}
+
+	response, err := m.client.Do(request)
+	if err != nil {
+		log.Warn().Err(err).Msg("Failed to request provider monitoring agent statuses")
+		return nil
```
Comment (soffokl) [inline core/quality/mysterium_morqa.go:304]:
```suggestion
		return err
```

File: core/quality/mysterium_morqa.go [hunk 4]
```diff
@@ -288,6 +289,32 @@ func (m *MysteriumMORQA) ProviderSessions(providerID string) []ProviderSession {
...
+
+	request, err := m.newRequestJSON(http.MethodGet, fmt.Sprintf("providers/statuses?provider_id=%s", providerID), "")
+	if err != nil {
+		log.Warn().Err(err).Msg("Failed to create provider monitoring agent statuses request")
+		return nil
+	}
+
+	response, err := m.client.Do(request)
+	if err != nil {
+		log.Warn().Err(err).Msg("Failed to request provider monitoring agent statuses")
```
Comment (soffokl) [inline core/quality/mysterium_morqa.go:303]:
```suggestion
		log.Error().Err(err).Msg("Failed to request provider monitoring agent statuses")
```

File: core/quality/mysterium_morqa.go [hunk 5]
```diff
@@ -288,6 +289,32 @@ func (m *MysteriumMORQA) ProviderSessions(providerID string) []ProviderSession {
 	return responseBody.Connects
 }
 
+// ProviderStatuses fetch provider connectivity statuses from prometheus (monitoring agent)
+func (m *MysteriumMORQA) ProviderStatuses(providerID string) node.MonitoringAgentStatuses {
+
+	request, err := m.newRequestJSON(http.MethodGet, fmt.Sprintf("providers/statuses?provider_id=%s", providerID), "")
+	if err != nil {
+		log.Warn().Err(err).Msg("Failed to create provider monitoring agent statuses request")
+		return nil
```
Comment (soffokl) [inline core/quality/mysterium_morqa.go:298]:
```suggestion
		return err
```

File: core/quality/mysterium_morqa.go [hunk 6]
```diff
@@ -288,6 +289,32 @@ func (m *MysteriumMORQA) ProviderSessions(providerID string) []ProviderSession {
 	return responseBody.Connects
 }
 
+// ProviderStatuses fetch provider connectivity statuses from prometheus (monitoring agent)
+func (m *MysteriumMORQA) ProviderStatuses(providerID string) node.MonitoringAgentStatuses {
+
+	request, err := m.newRequestJSON(http.MethodGet, fmt.Sprintf("providers/statuses?provider_id=%s", providerID), "")
+	if err != nil {
+		log.Warn().Err(err).Msg("Failed to create provider monitoring agent statuses request")
```
Comment (soffokl) [inline core/quality/mysterium_morqa.go:297]:
```suggestion
		log.Error().Err(err).Msg("Failed to create provider monitoring agent statuses request")
```

File: core/quality/mysterium_morqa.go [hunk 7]
```diff
@@ -288,6 +289,32 @@ func (m *MysteriumMORQA) ProviderSessions(providerID string) []ProviderSession {
...
+	if err != nil {
+		log.Warn().Err(err).Msg("Failed to request provider monitoring agent statuses")
+		return nil
+	}
+	defer response.Body.Close()
+
+	var statuses node.MonitoringAgentStatuses
+
+	if err = parseResponseJSON(response, &statuses); err != nil {
+		log.Warn().Err(err).Msg("Failed to parse provider monitoring agent statuses")
```
Comment (soffokl) [inline core/quality/mysterium_morqa.go:311]:
```suggestion
		log.Error().Err(err).Msg("Failed to parse provider monitoring agent statuses")
```

File: core/quality/mysterium_morqa.go [hunk 8]
```diff
@@ -288,6 +289,32 @@ func (m *MysteriumMORQA) ProviderSessions(providerID string) []ProviderSession {
...
+		return nil
```
Comment (soffokl) [inline core/quality/mysterium_morqa.go:312]:
```suggestion
		return err
```

File: tequilapi/contract/node.go [hunk 9]
```diff
@@ -26,3 +26,9 @@ import (
 type NodeStatusResponse struct {
 	Status node.MonitoringStatus `json:"status"`
 }
+
+// MonitoringAgentResponse a monitoring agent statuses reflects amount of connectivity statuses for each service_type and status
```
Comment (soffokl) [inline tequilapi/contract/node.go:30]:
```suggestion
// MonitoringAgentResponse reflects amount of connectivity statuses for each service_type.
```

File: tequilapi/endpoints/nat.go [hunk 10]
```diff
@@ -44,6 +44,10 @@ type nodeStatusProvider interface {
 	Status() node.MonitoringStatus
 }
 
+type nodeMonitoringAgent interface {
+	Statuses() node.MonitoringAgentStatuses
+}
+
```
Comment (soffokl) [inline tequilapi/endpoints/nat.go:50]: Why is this in the `nat.go` file?

**Review by soffokl — CHANGES_REQUESTED** [review]
The first commit is again called `initial commit`. 
`Initial commit` is created only once per repository and should never be used in the same repo again. 
If every PR will contain an `initial commit`, we will have 5331 `initial commits` in the `master`.

**Review by soffokl — APPROVED** [review]

File: core/quality/mysterium_morqa.go [hunk 11]
```diff
@@ -288,6 +289,32 @@ func (m *MysteriumMORQA) ProviderSessions(providerID string) []ProviderSession {
 	return responseBody.Connects
 }
 
+// ProviderStatuses fetch provider connectivity statuses from quality oracle.
+func (m *MysteriumMORQA) ProviderStatuses(providerID string) (node.MonitoringAgentStatuses, error) {
+
```
Comment (tomasmik) [inline core/quality/mysterium_morqa.go:294]: This is a useless empty line

File: core/quality/mysterium_morqa.go [hunk 12]
```diff
@@ -288,6 +289,32 @@ func (m *MysteriumMORQA) ProviderSessions(providerID string) []ProviderSession {
...
+	request, err := m.newRequestJSON(http.MethodGet, fmt.Sprintf("providers/statuses?provider_id=%s", providerID), "")
+	if err != nil {
+		log.Error().Err(err).Msg("Failed to create provider monitoring agent statuses request")
```
Comment (tomasmik) [inline core/quality/mysterium_morqa.go:297]: `log.Error().Err(err)` and `log.Err(err)` do the same thing, so not a lot reason to also add that `Error()`

File: tequilapi/endpoints/node.go [hunk 13]
```diff
@@ -49,14 +58,34 @@ func (ne *NodeEndpoint) NodeStatus(c *gin.Context) {
...
+// description: Node connectivity statuses as seen by monitoring agent
+// responses:
+//   200:
+//     description: Monitoring agent statuses ("success"/"cancelled"/"connect_drop/"connect_fail/"internet_fail)
+//     schema:
+//       "$ref": "#/definitions/MonitoringAgentResponse"
+func (ne *NodeEndpoint) MonitoringAgentStatuses(c *gin.Context) {
+	res, err := ne.nodeMonitoringAgent.Statuses()
+	if err != nil {
+		utils.WriteAsJSON(contract.MonitoringAgentResponse{Error: err.Error()}, c.Writer, http.StatusInternalServerError)
```
Comment (tomasmik) [inline tequilapi/endpoints/node.go:74]: you're missing a return here.

Comment (andrey-mysterium) [reply tequilapi/endpoints/node.go:74]: > you're missing a return here.

tanks!

**Review by tomasmik — APPROVED** [review]

**Pull request merged**